### DISCUSSION TITLE
Issue with transition from Plugin interface to Unified runtime

### DISCUSSION THREAD

**Discussion by tneval** [intel/llvm#15139] [Q&A] [answered]
As a context, I have been working with dpc++ offloading SYCL kernels to PoCL. The DPC++ version that I used still had plugin interfaces . Today, I fetched up-to-date dpc++, which is now using unified runtime. Rebuild succeeds, but it is somehow misbehaving. Here are some observations:

sycl-ls finds the device through PoCL:
`$ sycl-ls`
`[opencl:cpu][opencl:0] Portable Computing Language, cpu-znver3-AMD EPYC 7643 48-Core Processor OpenCL 3.0 PoCL HSTR: cpu-x86_64-unknown-linux-gnu-znver3`

Compiling sources with SYCL kernels succeeds (but offloading fails at PoCL's end) :
`$ clang -fsycl ex1.cpp`

`$ SYCL_DUMP_IMAGES=1 ./a.out`
`Selected device: cpu-znver3-AMD EPYC 7643 48-Core Processor
AuthenticAMD 
terminate called after throwing an instance of 'sycl::_V1::exception' 
 what():  Native API failed. Native API returns: 4 (UR_RESULT_ERROR_INVALID_VALUE)
Aborted (core dumped) `

Upon inspection, there is something off with produced SPIR-V.
Trying to disassemble:
`$ spirv-dis sycl_spir64.spv -o s.spirv`
`error: 8: Invalid capability operand: 5910`

Also, output of the trace:
```
$ SYCL_UR_TRACE=1 ./a.out

<LOADER>[INFO]: failed to load adapter 'libur_adapter_level_zero.so.0' with error: libze_loader.so.1: cannot open shared object file: No such file or directory
<LOADER>[INFO]: failed to load adapter '/home/tapio/PROJECT/Software/dpc++_T/lib/libur_adapter_level_zero.so.0' with error: libze_loader.so.1: cannot open shared object file: No such file or directory
<LOADER>[INFO]: loaded adapter 0x0x1dc93d0 (libur_adapter_opencl.so.0)
<LOADER>[INFO]: failed to load adapter 'libur_adapter_cuda.so.0' with error: libur_adapter_cuda.so.0: cannot open shared object file: No such file or directory
<LOADER>[INFO]: failed to load adapter '/home/tapio/PROJECT/Software/dpc++_T/lib/libur_adapter_cuda.so.0' with error: /home/tapio/PROJECT/Software/dpc++_T/lib/libur_adapter_cuda.so.0: cannot open shared object file: No such file or directory
<LOADER>[INFO]: failed to load adapter 'libur_adapter_hip.so.0' with error: libur_adapter_hip.so.0: cannot open shared object file: No such file or directory
<LOADER>[INFO]: failed to load adapter '/home/tapio/PROJECT/Software/dpc++_T/lib/libur_adapter_hip.so.0' with error: /home/tapio/PROJECT/Software/dpc++_T/lib/libur_adapter_hip.so.0: cannot open shared object file: No such file or directory
<LOADER>[INFO]: failed to load adapter 'libur_adapter_native_cpu.so.0' with error: libur_adapter_native_cpu.so.0: cannot open shared object file: No such file or directory
<LOADER>[INFO]: failed to load adapter '/home/tapio/PROJECT/Software/dpc++_T/lib/libur_adapter_native_cpu.so.0' with error: /home/tapio/PROJECT/Software/dpc++_T/lib/libur_adapter_native_cpu.so.0: cannot open shared object file: No such file or directory
SYCL_UR_TRACE: Requested device_type: info::device_type::cpu
SYCL_UR_TRACE: Selected device: -> final score = 2000
SYCL_UR_TRACE:   platform: Portable Computing Language
SYCL_UR_TRACE:   device: cpu-znver3-AMD EPYC 7643 48-Core Processor
Selected device: cpu-znver3-AMD EPYC 7643 48-Core Processor
AuthenticAMD
terminate called after throwing an instance of 'sycl::_V1::exception'
  what():  Native API failed. Native API returns: 4 (UR_RESULT_ERROR_INVALID_VALUE)
Aborted (core dumped)
```

Any ideas how to start dealing with this?

Thanks

**Comment by bader** [comment]
Please, try `SYCL_UR_TRACE=-1 ./a.out` (see
https://github.com/callumfare/llvm/blob/c854e3498309ab0501a1c6493bd239cf7f77ea65/sycl/doc/EnvironmentVariables.md#sycl_ur_trace-options for additional information about SYCL_UR_TRACE values). 

Hopefully this will give you more information about which UR and OpenCL calls are failing.

**Reply by tneval** [reply]
Here is the tracing of PoCL + DPC++:

```
|   GENERAL |  clGetDeviceInfo query of param 4250 for cpu
[2024-08-21 07:22:11.224009231] PoCL: in fn pocl_basic_get_device_info_ext at line 873:
|     ERROR |  Unknown param_name for get_device_info_ext: 16976
---> urDeviceGetInfo(.hDevice = 0x1d4b580, .propName = UR_DEVICE_INFO_IP_VERSION, .propSize = 4, .pPropValue = 0x7ffd941005f8 (2484078240), .pPropSizeRet = nullptr) -> UR_RESULT_ERROR_INVALID_VALUE;

terminate called after throwing an instance of 'sycl::_V1::exception'
what():  Native API failed. Native API returns: 4 (UR_RESULT_ERROR_INVALID_VALUE)
Aborted (core dumped)
```

So, it seems to be related to Intel extensions (0x4250).
I also rebuilt PoCL recently, so I will have to look if there is something strange.

Edit:
Previous build of DPC++, that I was using, did not request 0x4250.

**Reply by bader** [reply, marked as answer]
https://github.com/intel/llvm/issues/15149 seems to cover a similar problem. @aarongreig, @coldav, am I right?

**Reply by aarongreig** [reply]
yeah looks like the same problem, we have someone looking into it
